pkg/common/peertracker/getpid.c

```diff
@@ -0,0 +1,52 @@
+#include <windows.h>
+#include <iphlpapi.h>
+#include <stdio.h>
+
+unsigned long getOwningPIDFromLocalConn(int localPort, int remotePort, int *pid)
+{
+	PMIB_TCPTABLE_OWNER_PID pTcpTable = NULL;
+	DWORD dwSize = 0;
+	unsigned long retCode = 0;
+
+	for (;;) {
+		if (dwSize > 0) {
+			pTcpTable = (PMIB_TCPTABLE_OWNER_PID)calloc(dwSize, 1);
+			if (pTcpTable == NULL) {
+				return ERROR_OUTOFMEMORY;
+			}
+		}
+
+		retCode = GetExtendedTcpTable(pTcpTable, &dwSize, TRUE, AF_INET, TCP_TABLE_OWNER_PID_ALL, 0);
+		if (retCode == NO_ERROR) {
+			break;
+		}
+		if (retCode != ERROR_INSUFFICIENT_BUFFER) {
+			free(pTcpTable);
+			return retCode;
+		}
+	}
+
+	for (DWORD i = 0; i < pTcpTable->dwNumEntries; i++) {
+		if (pTcpTable->table[i].dwLocalAddr == htonl(INADDR_LOOPBACK) &&
+			pTcpTable->table[i].dwRemoteAddr == htonl(INADDR_LOOPBACK)  &&
+			ntohs((u_short)pTcpTable->table[i].dwLocalPort) == localPort &&
+			ntohs((u_short)pTcpTable->table[i].dwRemotePort) == remotePort &&
+			pTcpTable->table[i].dwState == MIB_TCP_STATE_ESTAB) {
+				*pid = pTcpTable->table[i].dwOwningPid;
+
+				if (pTcpTable != NULL) {
+					free(pTcpTable);
+					pTcpTable = NULL;
+				}

```

Comment:
nit: pTcpTable will certainly be non-nil here. Also, no need to set to NULL after the free since we're immediately returning. The pattern being used is generally useful for a generic cleanup routine or goto-based flow control, but here it just seems unnecessary to me.